src/videoencodingmonitor/VideoEncodingMonitor.ts

```diff
@@ -0,0 +1,173 @@
+    let videoInputFps: number = 0;
+    let videoEncodeFps: number = 0;
+    let isHardwareEncoder: boolean = false;
+
+    const attendeeId = this.context.meetingSessionConfiguration.credentials.attendeeId;
+
+    const metricData: { [id: string]: { [id: string]: {} } } = videoMetricReport[attendeeId];
+    const streams = metricData ? Object.keys(metricData) : [];
+    if (streams.length === 0) {
+      return;
+    }
+
+    for (const ssrc of streams) {
+      for (const [metricName, value] of Object.entries(metricData[ssrc])) {
+        switch (metricName) {
+          case 'videoUpstreamFramesInputPerSecond':
+            videoInputFps = Math.trunc(Number(value));
+            break;
+          case 'videoUpstreamFramesEncodedPerSecond':
+            videoEncodeFps = Math.trunc(Number(value));
+            break;
+          case 'videoUpstreamEncoderImplementation':
+            isHardwareEncoder = value as boolean;
+            break;
+        }
+      }
+    }
+
+    const hwEncoderFailed = isHardwareEncoder && videoEncodeFps === 0 && videoInputFps > 0;
+    if (hwEncoderFailed) {
+      this.continuousHwEncodeFailureCnt++;
+      if (
+        this.continuousHwEncodeFailureCnt >
+        VideoEncodingMonitor.CONTINUOUS_HW_ENCODE_FAILURE_THRESHOLD
+      ) {
+        this.degradeVideoCodec();
+        this.context.statsCollector.videoCodecDegradationHwEncodeFailureDidReceive();
+        this.continuousHwEncodeFailureCnt = 0;
+      }
+    } else {
+      this.continuousHwEncodeFailureCnt = 0;
+    }
+    return;
+  }
+
+  /**
+   * Degrade video codec to an alternative codec
+   */
+  private degradeVideoCodec(): void {
+    // Degrade video codec if there are other codec options and current codec is not H264 CBP or VP8
+    if (
+      this.context.meetingSupportedVideoSendCodecPreferences !== undefined &&
+      this.context.meetingSupportedVideoSendCodecPreferences.length > 1 &&
+      !(
+        this.context.meetingSupportedVideoSendCodecPreferences[0].equals(
+          VideoCodecCapability.h264ConstrainedBaselineProfile()
+        ) ||
+        this.context.meetingSupportedVideoSendCodecPreferences[0].equals(VideoCodecCapability.vp8())
+      )
+    ) {
+      const newMeetingSupportedVideoSendCodecPreferences: VideoCodecCapability[] = [];
+      for (const capability of this.context.videoSendCodecPreferences) {
+        if (!capability.equals(this.context.meetingSupportedVideoSendCodecPreferences[0])) {
+          newMeetingSupportedVideoSendCodecPreferences.push(capability);
+        }
+      }
+      if (newMeetingSupportedVideoSendCodecPreferences.length > 0) {
+        this.context.logger.info(
+          `Downgrading codec to ${newMeetingSupportedVideoSendCodecPreferences[0].codecName} due to slow encoding`
+        );
+        this.context.meetingSupportedVideoSendCodecPreferences = newMeetingSupportedVideoSendCodecPreferences;
+        this.context.audioVideoController.setVideoCodecSendPreferences(
```

Comment:
I think we should just call `update` instead? this will overwrite what the builder puts in, which we probably want to always know (e.g. in the case of recovery, etc).

Reply:
Just calling `update` will get `meetingSupportedVideoSendCodecPreferences` refreshed at next index frame. That can happen frequently. I added another array `videoSendCodecsBlocklisted` for blocklisted codecs to avoid constantly getting back to the originally selected codec while not overwriting builder input